client/src/infographic/PanelFilterControl.js

```diff
@@ -0,0 +1,164 @@
+  static getDerivedStateFromProps(nextProps, prevState) {
+    const potentially_new_table_tag_state = get_default_table_tag_state(
+      nextProps
+    );
+
+    const next_table_ids = _.keys(potentially_new_table_tag_state);
+    const prev_table_ids = _.keys(prevState.table_tags);
+    const table_set_has_changed =
+      next_table_ids.length !== prev_table_ids.length ||
+      _.some(
+        next_table_ids,
+        (table_id) => !_.includes(prev_table_ids, table_id)
+      );
+
+    if (table_set_has_changed) {
+      return {
+        table_tags: potentially_new_table_tag_state,
+      };
+    } else {
+      return null;
+    }
+  }
+  shouldComponentUpdate(nextProps, nextState) {
+    return this.state.table_tags !== nextState.table_tags;
+  }
+  componentDidUpdate() {
+    this.props.set_panel_filter(this.panel_filter_factory());
+  }
+  render() {
+    const { panel_keys } = this.props;
+    const { table_tags } = this.state;
+
+    const panel_filter = this.panel_filter_factory();
+
+    const tags = _.chain(table_tags)
+      .map((active, table_id) => ({
+        id: table_id,
+        label: Table.lookup(table_id).name,
+        active,
+      }))
+      .sortBy("label")
+      .value();
+
+    return (
+      <Details
+        summary_content={
+          <div>
+            <TM style={{ fontSize: 16 }} k="filter_panels" />
+            <TM
+              className="panel-status-text"
+              k="panels_status"
+              args={{
+                total_number_of_panels: panel_keys.length,
+                number_of_active_panels: panel_filter(panel_keys).length,
+              }}
+            />
+          </div>
+        }
+        persist_content={true}
+        content={
+          <LabeledBox
+            label={text_maker("filter_panels_description")}
+            children={
+              <div>
+                <TagCloud tags={tags} onSelectTag={this.onSelect} />
+                <div
+                  style={{
+                    borderTop: `1px dashed ${window.infobase_color_constants.tertiaryColor}`,
+                    padding: "10px 0px 10px 5px",
+                  }}
+                >
+                  <SelectAllControl
+                    SelectAllOnClick={this.onSelectAll}
+                    SelectNoneOnClick={this.onSelectNone}
+                  />
+                </div>
+              </div>
+            }
+          />
+        }
+      />
+    );
+  }
+  onSelect = (table_id) => {
+    const { table_tags } = this.state;
+
+    this.setState({
+      table_tags: { ...table_tags, [table_id]: !table_tags[table_id] },
+    });
+  };
+  onSelectAll = () => {
+    const { table_tags } = this.state;
+
+    this.setState({
+      table_tags: _.mapValues(table_tags, _.constant(true)),
+    });
+  };
+  onSelectNone = () => {
+    const { table_tags } = this.state;
+
+    this.setState({
+      table_tags: _.mapValues(table_tags, _.constant(false)),
+    });
+  };
+  panel_filter_factory = () => {
```

Comment:
I find this setup a bit confusing. It feels more natural to me to bring this into `Infographic` level, and just pass `number_of_active_panels` in as  props. Upto you on this change tho

Reply:
This approach was the most decoupled one I could think of. `Infographic.js` knows as little as it can about the implementation of `PanelFilterControl`.  Leaned hard in this direction since `Infographic.js` is already, IMO, doing too many things, and because it's a minority of infographics that actually get filters for now.

For `number_of_active_panels` specifically, if it was up to `Infographic.js` to calculate that then `Infographic.js` would need to know about the special case with "static" panels. There's no other reason for `Infographic.js` to be concerned if a panel is "static" or not, it's only a meaningful distinction when it comes to the implementation of the filter. Just good separation of concerns to move all that logic down in to `PanelFilterControl`.